Updates

hcmh opened the pull request:
Hi Martin,

here are some more updates. 

At least for my, the current master fails `make gputest` because of the error bound changed in the first commit.

The second commit is really just a bugfix, especially for `nlop_flatten`.

Then we have the optimization for linop_plus.

hcmh commented:
Ok, that's unexpected. Let me investigate where the travis failure first appears.

uecker commented:
I wonder whether we should simply check whether the lower-level forward operator is null?  Then the linear operator could stay as it is. Also better use CAST_MAYBE to do the type check rather then accessing the TYPEID directly. A cherry-picked two of the other commits. The gpu test works fine for me...

hcmh commented:
I have removed the gputest and have now changed it to use CAST_MAYBE. We still need to check whether we actually got a linop_data_t, so we still have two checks in the if() condition.

hcmh commented:
You are correct, now the second commit can be simplified.
